java-frontend/src/main/java/org/sonar/java/resolve/Resolve.java

```diff
@@ -440,12 +440,15 @@ private Resolution selectBest(Env env, JavaType site, List<JavaType> argTypes, L
     if(!hasCompatibleArity(methodJavaSymbol.parameterTypes().size(), argTypes.size(), methodJavaSymbol.isVarArgs())) {
       return bestSoFar;
     }
-    TypeSubstitution substitution = typeSubstitutionSolver.getTypeSubstitution(methodJavaSymbol, site, typeParams, argTypes);
+    TypeSubstitution substitution = typeSubstitutionSolver.getTypeSubstitution(methodJavaSymbol, callSite, typeParams, argTypes);
     if (substitution == null) {
       return bestSoFar;
     }
     List<JavaType> formals = ((JavaType.MethodJavaType) methodJavaSymbol.type).argTypes;
-    formals = typeSubstitutionSolver.applySiteSubstitutionToFormalParameters(formals, site);
+    formals = typeSubstitutionSolver.applySiteSubstitutionToFormalParameters(formals, callSite);
+    if(defSite != callSite) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Change this comparison to use the equals method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1698)